File: leetcode_py/Permutations.py
class Solution:
    def next(self, cur):
        for j in range(len(cur) - 2, -1, -1):
            for i in range(len(cur) - 1, j, -1):
                if cur[i] > cur[j]:
                    cur[i], cur[j] = cur[j], cur[i]
                    cur[j + 1:] = cur[j + 1:][::-1]
                    return cur
        return False

    def permute(self, num):
        num = sorted(num)
        res = [num[::]]
        while num:
            num = self.next(num)
            if num:
                res.append(num[::])
        return res


class Solution:
    def dfs(self, num, path, res, flag):
        if len(path) == len(num):
            res.append(path[:])
            return
        for i in range(len(num)):
            if not flag[i]:
                flag[i] = True
                path.append(num[i])
                self.dfs(num, path, res, flag)
                flag[i] = False
                path.pop()

    def permute(self, num):
        res, flag = [], [False for i in num]
        self.dfs(num, [], res, flag)
        return res


if __name__ == "__main__":
    a = Solution()

    print a.permute([2, 34])
    print a.permute([3, 1, 2])
    print a.permute([1, 2, 3, 4])

File: leetcode_js/Permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
   var res = [], flag = [], path = [];
   for(var i = 0; i < nums.length; i++){flag.push(false);}
   function dfs(){
   		if(path.length == nums.length){
   			res.push(path.slice());
   			return;
   		}
   		for(var i = 0; i < nums.length; i++){
   			if(!flag[i]){
   				flag[i] = true;
   				path.push(nums[i]);
   				dfs();
   				flag[i] = false;
   				path.pop();
   			}
   		}
   } 
   dfs();
   return res;
};


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    nums = nums.sort(function(a, b){
    	return a - b;
    }); 
    var res = [];
    while(nums){
    	res.push(nums.slice(0));
    	nums = next(nums); 
    }
    return res;
};
function next(cur){
	for(var j = cur.length - 2; j > -1; j--){
		for(var i = cur.length - 1; i > j; i--){
			if(cur[i] > cur[j]){
				var tmp = cur[j];
				cur[j] = cur[i];
				cur[i] = tmp;
				var k = j + 1, m = cur.length - 1;
				while(k < m){
					var tmp = cur[k];
					cur[k++] = cur[m];
					cur[m--] = tmp;
				}  
				return cur;
			}
		}
	}
	return false;
}